### DISCUSSION TITLE
Conditional debugging

### DISCUSSION THREAD

**Discussion by arduano** [SeaQL/FireDBG.for.Rust#11] [Ideas]
Hi! I'm really loving this project so far, and I hope I didn't just miss something in the documentation about this but:

I often use Rust for very high-performance real-time (e.g. game/graphics/audio stuff) applications, often multithreaded, which is of course difficult to debug. This project seems like it has the potential to make my life much much easier, but currently it seems to collect everything, which makes it infeasible to debug the kinds of applications I make, both due to the performance hit and the amount of data it collects.

Basically what I'm proposing is, would it be possible to make a macro/wrapper that can conditionally enable debugging for all child function calls? e.g. I don't want to slow down my graphics when debugging a single function in my audio system. And in the audio system, I don't want to slow down the whole audio system when I'm debugging one edge case with known inputs that cause it.

But yeah, looking forward to the future of this project!

**Comment by astrale-sharp** [comment]
I think what I'd like resembles that:
I can't find a way to have a breakpoint other than by panicking at the moment, I think we could maybe have a
`fire::breakpoint!()`. that would be really easy to setup
We could hide it behind an if condition to get what @arduano is asking!

**Reply by tyt2y3** [reply]
Actually, you can use the `fire::dbg!` macro, it sets a breakpoint in place. Although right now it only captures the variable being passed in, but it's possible to extend it to do some more operations.

**Comment by astrale-sharp** [comment]
Also would like to double down on the enthousiasm I share for this project, the state of rust debugging is globally bad, hard to setup etc and I think this is very promising !

**Reply by tyt2y3** [reply]
Yeah, I mean we have gone by without a first-party debugger for years and a lot of work is needed to catch up to the complexity of the Rust language today.

**Comment by tyt2y3** [comment]
Thank you for starting a discussion. Replying to OP, `enable debugging for all child function calls`: yes, what I have been thinking is to put macros to annotate different debugging profiles, something like:

```rust
#[firedbg(profile = "audio")]
```

and so you can precisely filter which parts of the program to trace by selecting the right profile.

But a big blocker for now is we don't have the static call-tree, so we can't intelligently filter the children. I think the most doable way is to fork and trim down rust analyzer to only do enough work to extract the static call-tree, and to dump that info to disk.
